### ISSUE TITLE
AdMob banner ads not showing after app pause/resume

### ISSUE THREAD

**Issue by faustob** [googleads/googleads-mobile-unity#2970]
### [REQUIRED] Step 1: Describe your environment

  * Unity version: 2022.3.10f1
  * Google Mobile Ads Unity plugin version: 8.5.2
  * Platform: Android
  * Platform OS version: Android 13
  * Any specific devices issue occurs on: All Android devices tested
  * Mediation ad networks used, and their versions: Only Admob

### [REQUIRED] Step 2: Describe the problem

Banner ads show correctly at app start, and as long as the app stays in focus. If users switch to a different app, or the phone locks, on app resume the banner ads are no longer visible. This only happens on Android, the same app on iOS works fine. It happens in both development and release builds.

@NVentimiglia , I see your workaround [here](https://github.com/googleads/googleads-mobile-unity/issues/1342#issuecomment-1433924368), but I am not sure how to implement it. I tried creating a Unity plugin following https://docs.unity3d.com/Manual/android-custom-activity.html, but could not get it to work.

Is the workaround still valid?
Can you provide instructions?
Is this bug being addressed?

[ FYI, I have also posted this issue as a [comment](https://github.com/googleads/googleads-mobile-unity/issues/1342#issuecomment-1752089134) to the previous (closed) thread ]

#### Steps to reproduce:

Happy to provide logs or other info, just not sure what would help.
Are others seeing this behavior?

**Comment by viveksaini75** [comment]
I am also facing the same issue

**Comment by nguyendai2193** [comment]
same issue, please help

**Comment by VectorViral** [comment]
Same issue ,Loots of bugs in new SDK after updating SDK from 7... to 8.5.2

**Comment by Untumuxx** [comment]
same issue

**Comment by VectorViral** [comment]
Admob is turning into trash day by day....

**Comment by faustob** [comment]
@NVentimiglia is there an update or workaround for this issue?
It seems pretty important to me, leading to significant loss of ad revenue.

Thanks!

**Comment by faustob** [comment]
I tried the following:

1) located file in /Applications/Unity/PlaybackEngines/AndroidPlayer/Source/com/unity3d/player/UnityPlayerActivity.java

2) replaced onResume with the code you provided in https://github.com/googleads/googleads-mobile-unity/issues/1342#issuecomment-1433924368

```
@Override
  protected void onResume() {
    super.onResume();
    mUnityPlayer.resume();
    KeyguardManager keyManager = (KeyguardManager) this.getSystemService(Context.KEYGUARD_SERVICE);
    KeyguardDismissCallback keyguardDismissCallback =
        new KeyguardDismissCallback() {
          @Override
          public void onDismissSucceeded() {}

          @Override
          public void onDismissError() {}

          @Override
          public void onDismissCancelled() {}
        };
    keyManager.requestDismissKeyguard(this, keyguardDismissCallback);
  }
```

3) quit Unity Editor and re-opened project
4) built for Android
5) Tested on Android device. Same problem as before.

I would appreciate any additional pointers!

**Comment by faustob** [comment]
I figured out (I think) how to implement the workaround from https://github.com/googleads/googleads-mobile-unity/issues/1342#issuecomment-1433924368

However, the behavior did not change. Banners still disappear after an app pause/resume (eg locking the screen and returning, or switching to another app and then back).

If anybody wants to try by themselves, here's what I did:

1) Locate UnityPlayerActivity.java in your Unity install. On my Mac, this is the path:
/Applications/Unity/Hub/Editor/2022.3.11f1/PlaybackEngines/AndroidPlayer/Source/com/unity3d/player/UnityPlayerActivity.java

2) Replace onResume() with this code:

```
import android.app.KeyguardManager;
import android.app.KeyguardManager.KeyguardDismissCallback;
import android.content.Context;

//...

    @Override
    protected void onResume() {
        super.onResume();
        mUnityPlayer.resume();
        KeyguardManager keyManager = (KeyguardManager) this.getSystemService(Context.KEYGUARD_SERVICE);
        KeyguardDismissCallback keyguardDismissCallback = new KeyguardDismissCallback() {
            @Override
            public void onDismissSucceeded() {
            }

            @Override
            public void onDismissError() {
            }

            @Override
            public void onDismissCancelled() {
            }
        };
        keyManager.requestDismissKeyguard(this, keyguardDismissCallback);
    }
```

3) Build for Android and test on device.

Hopefully somebody can figure out how to take it from here...

**Comment by NVentimiglia** [comment]
@faustob 

Thanks for your patience, I was out.

First, what are the replication steps for this issue. Here is what I tried 👍 

1.  Start with the hello world sample. 
2. Load a banner ad.
3. Hide the banner ad.
4. Background the ap.
5. Press the power button.
6. Wait 2 minutes.
7. Press power button and return to app.
8. Show the ad.

When doing this, the banner ad showed as expected.


Second, [The android custom activity](https://docs.unity3d.com/Manual/android-custom-activity.html) is a pretty complicated fix. I could try pushing for a fix using `KeyguardManager`, but I am concerned with unintended consequences.  Could you implement a fix in C# using `OnApplicaitonPause()` ?

**Comment by IoannisFMG** [comment]
We have the same issue, but we are not using the Unity AdmobSDK. Rather we use our own implementation.
I think this might have something to do with a Unity Player change, rather than an Admob SDK one. 

I have some findings to share.

I tested it on a Pixel (OS 13) and I noticed that if you hit home and return to the app from the device's "task manager" the app behaves normally. 
If the app moves to the background fully (gets completely minimized) then the problem starts to happen. 
It also happens some times after seeing an interstitial. Didn't always happen, so I guess it might have to do with the mediation that served the ad.

**Comment by NVentimiglia** [comment]
@IoannisFMG if you have specific replication steps I can take another look.

**Comment by IoannisFMG** [comment]
@NVentimiglia, I'll see if I can reproduce it in a simple test project and I'll let you know. Unfortunately, I can't share company code.

**Comment by lucas-brasmobi** [comment]
same issue

**Comment by faustob** [comment]
@IoannisFMG Thanks for your efforts! Did you submit a Unity bug?

**Comment by IoannisFMG** [comment]
Not yet, because I don't know if this thing needs to be a common effort between the Admob Team and Unity.

I don't know if Unity will be able to handle this without having some insight on what is not being called, or what is called and shouldn't be etc.

I could create a ticket if @NVentimiglia has no idea why the issue happens.

**Comment by emathis11** [comment]
Downgrading from Unity 2022.3.10 to 2022.3.6 fixes the problem for us.

**Comment by faustob** [comment]
Thank you @emathis11 , that worked for me as well!

I will submit a bug to Unity. @IoannisFMG, may I submit the files you provided to reproduce the bug? (Alternatively, do you want to submit the bug?)

**Comment by jeffyuio1129** [comment]
The same problem occurs. Is there a solution?

**Comment by faustob** [comment]
> The same problem occurs. Is there a solution?

What do you mean? Did you try to downgrade to 2022.3.6, as suggested above? Did that not work?

**Comment by viveksaini75** [comment]
Bro, I have been using 2021.3.21f1 for the last 7 months, and when upgraded to 8.5.2 I faced the same issue.

**Comment by viveksaini75** [comment]
I think it's not an editor bug

**Comment by IoannisFMG** [comment]
Yes, @faustob, I will submit one now. I'll share the link once it's public.

**Comment by jeffyuio1129** [comment]
> > 같은 문제가 발생합니다. 해결책이 있나요?
> 
> 무슨 뜻이에요? 위에서 제안한 대로 2022.3.6으로 다운그레이드를 시도하셨나요? 작동하지 않았나요?

downgrade is not possible due to other issues.

is there no solution in 2022.3.12f1?

**Comment by NVentimiglia** [comment]
@IoannisFMG 

The attached video did not download / work. 

I am still unable to replicate the issue.

If you have a sample video and sample code, submit it to [The google support group](https://groups.google.com/g/google-admob-ads-sdk).

**Comment by IoannisFMG** [comment]
@NVentimiglia, here https://www.youtube.com/shorts/dR3WYKV6h6w

**Comment by danhtran-dev** [comment]
I think this is an issue of Unity Renderer

This is the Layout inspector before pause Unity Game Activity. Can see the Surface is behide the Banner.
![Screenshot 2023-11-02 at 01 16 33](https://github.com/googleads/googleads-mobile-unity/assets/67886170/5e13b926-15c9-4da7-9f85-880951ff6f14)


This is the layout inspector after we resume. Can see, The Unity Surface covered Banner View
![Screenshot 2023-11-02 at 01 13 57](https://github.com/googleads/googleads-mobile-unity/assets/67886170/e21dc077-39e9-40c3-bf7a-74aa684772a5)

**Comment by faustob** [comment]
> @IoannisFMG
> 
> The attached video did not download / work.
> 
> I am still unable to replicate the issue.
> 
> If you have a sample video and sample code, submit it to [The google support group](https://groups.google.com/g/google-admob-ads-sdk).

FYI - First time I tried to download I missed that I had to "Accept terms and conditions" on the page to access the file.

**Comment by NVentimiglia** [comment]
@faustob 

Ty for your feedback, I will talk with that team to see if we can streamline file uploading.

**Comment by NVentimiglia** [comment]
Internal bug id for reference b/308855324

The fact that Unity pushes the banner behind the game is a problem worth looking into.

**Comment by faustob** [comment]
> @faustob
> 
> Ty for your feedback, I will talk with that team to see if we can streamline file uploading.

I meant to say that I had that problem when trying to download the files created by @IoannisFMG  from [this link](https://we.tl/t-JehKUBAK0U). He included a video as well as sample code to replicate the issue. Thought you may have had the same issue. Sorry for the confusion.

**Comment by danhtran-dev** [comment]
Hey bros, In the mean time what is a work around solution here. Down grade Unity to a particular version?

**Comment by itamishi** [comment]
The only "workaround" I came up with is to create a new banner every time the app unpauses. 

I'm NOT SURE if it's a violation of AdMob policy, as it probably counts as a separate banner show

```
    private void OnApplicationPause(bool pause) {
#if !UNITY_EDITOR && UNITY_ANDROID
        if (!pause) {
            Banner.Instance.CreateBannerView();
        }
#endif
    }
```

Banner code: (its a Singleton)

```
    public void CreateBannerView() {
        Debug.Log("Creating banner view");

        if (_bannerView != null) {
            DestroyBannerView();
        }

        _bannerView = new BannerView(_adUnitId, AdSize.Banner, AdPosition.Bottom);

        var adRequest = new AdRequest();
        _bannerView.LoadAd(adRequest);
    }
```

I think of implementing some sort of timer (of 1 minute for example) so the user can't spam minimize/maximize app, to make it a bit safer

**Comment by faustob** [comment]
For me all it took was downgrading to Unity 2022.3.6, as suggested above (newer versions have the issue.) That version has an issue compiling for iOS, so I now switch between the two version to build for Android and iOS :(

**Comment by faustob** [comment]
@IoannisFMG any update on bug submission? Thanks!

**Comment by IoannisFMG** [comment]
Unfortunately no updates, @faustob. The ticket I submitted is still open. They haven't even looked at it, yet.

**Comment by ld07** [comment]
Same problem here... But Unity version 2021.3.32.  My workaround was downgrading to Unity 2021.3.29. It seems that the problem arose as a side effect of this fix that was made in both Unity 2021.3.30 and 2022.3.6:

![image](https://github.com/googleads/googleads-mobile-unity/assets/26928558/64bbb5df-bed7-4e0b-bda7-43dad0e46edd)


"Android: Fixed an issue where pause was not processed when onPause() is called right after onResume() is called. ([UUM-23247](https://issuetracker.unity3d.com/issues/button-cant-be-pressed-after-calling-startactivityforresult-with-flag-activity-new-task))"

**Comment by IoannisFMG** [comment]
By the way, the Incident number I reported is `IN-59336`. In case needed somewhere.

**Comment by NVentimiglia** [comment]
Closing this issue. Please reopen if Unity gets back to you and has next steps for us.

**Comment by hasnain-shukat** [comment]
Happening in Unity 2022.3.12 too.

**Comment by shefich** [comment]
Happening in Unity 2022.3.13 and 2022.3.14.

**Comment by IoannisFMG** [comment]
Yesterday, they put the issue I submitted in Review. I'll keep you posted.

**Comment by IoannisFMG** [comment]
The bug has been confirmed and it has been forwarded to Unity's dev team. 
For reference, the internal ID is `UUM-57151`.

I'll post any updated I receive here.

**Comment by povilasxd3** [comment]
Hi everyone!
The issue has now been processed and you can follow its progress in the Issue Tracker here: https://issuetracker.unity3d.com/product/unity/issues/guid/UUM-57151
Thank you all for your patience

**Comment by oraylan** [comment]
I'm having the same problem here, apparently different people are having it. But I already saw that they opened a ticket, let's wait. Or does anyone already have a solution?

**Comment by Unarmed1000** [comment]
Given that the "New Consent Management Platform requirements" take effect January 16, 2024. It would seem critical to get a fix out asap so people have a chance of doing a release before then.

**Comment by Unarmed1000** [comment]
@NVentimiglia given the upcoming deadline imposed by google that requires apps to be updated before Jan16, would it not make sense to take charge of this and work with unity to solve this asap so all the Unity apps can be updated in time?

**Comment by IoannisFMG** [comment]
> A fix for your bug report IN-59336: [Android] AdmobSDK banner ad disappears when the device goes to the home screen and returns to the application has been submitted with the following status: Fixed and currently awaits internal confirmation.
>
>Fix Version/s: Pending

A fix is probably coming soon to an Editor version near you.

**Comment by faustob** [comment]
FYI, from https://issuetracker.unity3d.com/issues/android-admobsdk-banner-ad-disappears-when-the-device-goes-to-the-home-screen-and-returns-to-the-application

Under Consideration for 2021.3.X, 2024.1.X

Fixed in 2022.3.16f1, 2023.3.0a19

In Progress in 2023.2.X

Waiting for 2022.3.16f1 to be released...

**Comment by faustob** [comment]
2022.3.16f1 was released and based on my initial testing the bug is finally fixed.

As far as other editor versions, current state is:

Under Consideration for 2021.3.X, 2024.1.X
Fixed in 2022.3.16f1, 2023.3.0a19
In Progress in 2023.2.X

**Comment by mdolzhansky** [comment]
I upgrated unity to 2021.3.36f1 and now it fixed
